Sources/NIOSOCKS/Messages/SOCKSRequest.swift

```diff
@@ -0,0 +1,200 @@
+        var written = self.writeInteger(request.version)
+        written += self.writeInteger(request.command.value)
+        written += self.writeInteger(UInt8(0))
+        written += self.writeAddressType(request.addressType)
+        return written
+    }
+    
+}
+
+// MARK: - SOCKSCommand
+
+/// What type of connection the SOCKS server should establish with
+/// the target host.
+public struct SOCKSCommand: Hashable {
+    
+    /// Typically the primary connection type, suitable for HTTP.
+    public static let connect = SOCKSCommand(value: 0x01)
+    
+    /// Used in protocols that require the client to accept connections
+    /// from the server, e.g. FTP.
+    public static let bind = SOCKSCommand(value: 0x02)
+    
+    /// Used to establish an association within the UDP relay process to
+    /// handle UDP datagrams.
+    public static let udpAssociate = SOCKSCommand(value: 0x03)
+    
+    public var value: UInt8
+    
+    public init(value: UInt8) {
+        self.value = value
+    }
+}
+
+// MARK: - SOCKSAddress
+
+/// The address used to connect to the target host.
+public enum SOCKSAddress: Hashable {
+    
+    case address(SocketAddress)
+    
+    case domain(String, port: Int)
+    
+    static let ipv4IdentifierByte: UInt8 = 0x01
+    static let domainIdentifierByte: UInt8 = 0x03
+    static let ipv6IdentifierByte: UInt8 = 0x04
+    
+    /// How many bytes are needed to represent the address, excluding the port
+    var size: Int {
+        switch self {
+        case .address(.v4):
+            return 4
+        case .address(.v6):
+            return 16
+        case .address(.unixDomainSocket):
+            fatalError("Unsupported")
+        case .domain(let domain, port: _):
+            // the +1 is for the leading "count" byte
+            // containing how many UTF8 bytes are in the
+            // domain
+            return domain.utf8.count + 1
+        }
+    }
+}
+
+extension ByteBuffer {
+    
+    mutating func readAddressType() throws -> SOCKSAddress? {
+        return try self.parseUnwindingIfNeeded { buffer in
+            guard let type = buffer.readInteger(as: UInt8.self) else {
+                return nil
+            }
+            
+            switch type {
+            case SOCKSAddress.ipv4IdentifierByte:
+                return try buffer.readIPv4Address()
+            case SOCKSAddress.domainIdentifierByte:
+                return buffer.readDomain()
+            case SOCKSAddress.ipv6IdentifierByte:
+                return try buffer.readIPv6Address()
+            default:
+                throw SOCKSError.InvalidAddressType(actual: type)
+            }
+        }
+    }
+    
+    mutating func readIPv4Address() throws -> SOCKSAddress? {
+        return try self.parseUnwindingIfNeeded { buffer in
+            guard
+                let bytes = buffer.readSlice(length: 4),
+                let port = buffer.readPort()
```

Comment:
When we read short and return `nil`, where are we returning the read bytes to the buffer (i.e if we read `byte` but not `port`)?

Reply:
`ByteBuffer.parseUnwindingIfNeeded`

Reply:
That only restore when you `throw`, not when you return `nil`?

Reply:
Oh shit, this got killed in a bad merge, one second - I'll re-apply the commit